src/array.cr

```diff
@@ -1058,13 +1058,19 @@ class Array(T)
   end
 
   # Optimized version of `Enumerable#map_with_index`.
-  def map_with_index(&block : T, Int32 -> U) forall U
-    Array(U).new(size) { |i| yield @buffer[i], i }
+  #
+  # Accepts an optional *offset* parameter, which tells it to start counting
+  # from there.
+  def map_with_index(offset = 0, &block : T, Int32 -> U) forall U
+    Array(U).new(size) { |i| yield @buffer[i], offset + i }
```

Comment:
I dont really know the difference here but why not use `build`

```suggestion
    Array(U).build(size) { |i| yield @buffer[i], offset + i }
```

Reply:
Please refer to the documentation of [Array.build](https://crystal-lang.org/api/0.31.1/Array.html#build(capacity:Int,&block):self-class-method) vs. [Array.new](https://crystal-lang.org/api/0.31.1/Array.html#new(size:Int,&block:Int32-%3ET)-class-method).